src/vscode-dts/vscode.proposed.debugDataBreakpoints.d.ts

```diff
@@ -0,0 +1,162 @@
+		/** Reference to the stack frame to which the expression is scoped. */
+		frameId: number;
+		/** The name of the expression that is used for resolution. */
+		expression: string;
+		/** A human-readable label for the data breakpoint source. */
+		description: string;
+
+		/**
+		 * Creates a new frame scoped data breakpoint source.
+		 *
+		 * @param frameId Reference to the stack frame to which the expression is scoped.
+		 * @param expression The name of the expression that is used for resolution.
+		 * @param description A human-readable label for the data breakpoint source.
+		 */
+		constructor(frameId: number, expression: string, description?: string);
+	}
+
+	/** The source for a data breakpoint. */
+	export type DataBreakpointSource = ResolvedDataBreakpointSource | AddressDataBreakpointSource | ExpressionDataBreakpointSource | VariableScopedDataBreakpointSource | FrameScopedDataBreakpointSource;
+
+	/** Access type for data breakpoints. */
+	export type DataBreakpointAccessType = 'read' | 'write' | 'readWrite';
+
+	/**
+	 * A breakpoint specified by a variable or memory change.
+	 */
+	export class DataBreakpoint extends Breakpoint {
+		/** The source for the data breakpoint. See the different source types on how they are resolved during breakpoint activation. */
+		source: DataBreakpointSource;
+
+		/** The resolution of the data breakpoint based on the given source. This may be undefined if the breakpoint was not yet resolved. */
+		resolution?: DataBreakpointResolution;
+
+		/** The access type of the data. */
+		accessType: DataBreakpointAccessType;
+
+		/**
+		 * Create a new data breakpoint.
+		 *
+		 * @param source The source for the data breakpoint. If the `dataId` is known already, it can be specified directly. If the dataId is not known, it can be retrieved via a data breakpoint info request from the session. Alternatively, some sources offer dynamic resolution during breakpoint activation.
+		 * @param accessType The access type of the data breakpoint.
+		 * @param enabled Is breakpoint enabled.
+		 * @param condition Expression for conditional breakpoints.
+		 * @param hitCondition Expression that controls how many hits of the breakpoint are ignored.
+		 * @param logMessage Log message to display when breakpoint is hit.
+		 */
+		constructor(source: DataBreakpointSource | string, accessType: DataBreakpointAccessType, enabled?: boolean, condition?: string, hitCondition?: string, logMessage?: string);
+
+		/**
+		 * Resolves the data breakpoint and retrieves the dataId, canPersist, and description properties.
+		 * By default, the data breakpoint is resolved automatically after the breakpoint is set.
+		 * A data breakpoint is only resolved once at a certain point in time.
+		 * Calling this method again will not update the resolution properties properties.
+		 *
+		 * @param session the session against which this breakpoint should be resolved.
+		 */
+		resolve(session: DebugSession): Thenable<this>;
```

Comment:
Breakpoints are global and visible to all extensions. This changes the data on the DataBreakpoint to view what its resolution state is for a session, but if this is called concurrently between two different extensions or for two different sessions, you would get racey behavior. Instead of mutating DataBreakpoint, should this return the resolution on that session?